Support multi-user undo and redo

hyemmie opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it?
It's the initial version of multi-user undo/redo implementation. For now, it's focusing on implementing an architecture to support undo/redo using `liveblocks` as a reference. 

#### Any background context you want to provide?
Additional information about multi-user undo/redo is at https://github.com/yorkie-team/yorkie/issues/49.

#### What are the relevant tickets?
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

### Checklist
- [ ] Added relevant tests or not required
- [ ] Didn't break anything

hyemmie committed:
add TODO requirement comments

hyemmie committed:
add test code for initial undo/redo

Co-authored-by: Yourim Cha <yourim.cha@navercorp.com>

hyemmie committed:
add history at document

Co-authored-by: Yourim Cha <yourim.cha@navercorp.com>

hyemmie committed:
feat reverse operation for counter increase

Co-authored-by: Yourim Cha <yourim.cha@navercorp.com>

hyemmie committed:
feat undo function

hyemmie committed:
temp for test

hyemmie committed:
Fix update() function

hyemmie committed:
Add TODO comment for undefined executedAt

hyemmie committed:
Feat redo function

hyemmie committed:
Add redo and no change test cases

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/yorkie-team/yorkie-js-sdk?pullRequest=635) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/yorkie-team/yorkie-js-sdk?pullRequest=635) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/yorkie-team/yorkie-js-sdk?pullRequest=635) it.</sub>

hyemmie committed:
Add undo/redo error handling and test code

chacha912 committed:
Add assertUndoRedo and handle undo/redo for Long type

hyemmie committed:
Add max stack size for undo/redo and test code

hyemmie committed:
Merge branch 'feat/undo-redo-arch' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

hyemmie committed:
Fix long counter multiply function type error

chacha912 committed:
Add test for reverse operation of increase operation

hyemmie committed:
Add redo stack clear function and test code

chacha912 committed:
Add test for changeInfo when undoing increase operation

chacha912 committed:
Merge branch 'feat/undo-redo-arch' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

chacha912 committed:
Implement undo/redo with presence

chacha912 committed:
Merge branch 'main' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

hyemmie committed:
Feat reverse operation of object set, remove

hyemmie committed:
Update test code for object undo/redo

chacha912 committed:
Add codemirror devtool

hyemmie committed:
Add concurrent object undo/redo tests

hyemmie committed:
Fix set reverse operation bug

hyemmie committed:
Add assertUndoRedo to object_test

hyemmie committed:
Update document_test

chacha912 committed:
Merge branch 'feat/undo-redo-arch' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

chacha912 committed:
Add reverse operation of text.edit

chacha912 committed:
Add undo/redo in codemirror example

chacha912 committed:
Add test cases that fail to share for debugging

chacha912 committed:
Cleanup test code

hyemmie committed:
Implement the reverse operation of text.edit

hyemmie committed:
Update protocol buffer

hyemmie committed:
Add concurrent test cases of text.edit undo

chacha912 committed:
Cleanup undo/redo test for simple object and nested object

chacha912 committed:
Modify to execute reverse operations in reverse order within change

chacha912 committed:
Modify to enable creating CRDTElement with same createdAt

Currently, executedAt is used as createdAt, preventing the creation
of elements with same createdAt. To enable creating elements with
same createdAt during undo, this commit introduced lastUpdatedAt
and compared it to executedAt.

chacha912 committed:
Merge branch 'feat/undo-redo-arch' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

chacha912 committed:
Merge branch 'main' of https://github.com/yorkie-team/yorkie-js-sdk into feat/undo-redo-arch

chacha912 committed:
Add undo and redo shortcuts to CodeMirror example

hackerwins commented:
This continued in #650.
